Sqlite: implement withLock from DBLayer

rvl opened the pull request:
Relates to issue #154.

# Overview

This uses an MVar for application-level locking.

It _may_ be possible to implement using SQLite `BEGIN EXCLUSIVE TRANSACTION` but I would prefer to try and refine the `DBLayer` interface first. Maybe we can remove `withLock`.

paweljakubas commented on test/integration/Test/Integration/Scenario/Transactions.hs:
```diff
@@ -64,6 +63,44 @@ spec = do
             [ expectWalletError (UtxoNotEnoughFragmented (Client.ErrUtxoNotEnoughFragmented 1 Client.msgUtxoNotEnoughFragmented))
             ]
 
+    scenario "cannot send subsequenct transaction when the first one is pending" $ do
```
should be probably `subsequent`

rvl committed:
Sqlite: implement withLock from DBLayer

Uses an MVar for application-level locking.

This may be possible to implement using SQLite BEGIN EXCLUSIVE
TRANSACTION but I would prefer to try and refine the DBLayer interface
first.
